Composer/packages/server/src/services/project.ts

```diff
@@ -432,6 +436,126 @@ export class BotProjectService {
     }
   };
 
+  public static async migrateProjectAsync(req: Request, jobId: string) {
+    const { oldProjectId, name, description, location, storageId } = req.body;
+    // todo: use this
+    description;
+    // console.log(`Migrate ${oldProjectId} to ${location}/${name} with ${description}`);
+    const user = await ExtensionContext.getUserFromRequest(req);
+
+    try {
+      const locationRef = getLocationRef(location, storageId, name);
+
+      await BotProjectService.cleanProject(locationRef);
+
+      // Update status for polling
+      BackgroundProcessManager.updateProcess(jobId, 202, formatMessage('Getting template'));
+
+      const newProjRef = await AssetService.manager.copyRemoteProjectTemplateToV2(
+        '@microsoft/generator-microsoft-bot-adaptive',
+        '1.0.0-preview-20210310.8ee9434', // use any available version
+        name,
+        locationRef,
+        jobId,
+        user
+      );
+
+      // update project ref to point at newly created folder
+      newProjRef.path = `${newProjRef.path}/${name}`;
+
+      BackgroundProcessManager.updateProcess(jobId, 202, formatMessage('Migrating data'));
+
+      const originalProject = await BotProjectService.getProjectById(oldProjectId, user);
+      if (originalProject.settings) {
+        const originalFiles = originalProject.getProject().files;
+
+        // pass in allowPartialBots = true so that this project can be opened even though
+        // it doesn't yet have a root dialog...
+        const id = await BotProjectService.openProject(newProjRef, user, true);
+        const currentProject = await BotProjectService.getProjectById(id, user);
+
+        // add all original files to new project
+        for (let f = 0; f < originalFiles.length; f++) {
+          await currentProject.migrateFile(
+            originalFiles[f].name,
+            originalFiles[f].content,
+            originalProject.rootDialogId
+          );
+        }
+
+        const newSettings = {
```

Comment:
@carlosscastro can you look at this settings transform? 

I think should we whip up a json-schema definition and pull in a runtime validator and announce failures for the cases that users have modified settings into something we don't expect.

Reply:
That is a great idea!